server/protocol.c

```diff
@@ -1466,22 +1120,196 @@ static void apply_server_config(request_rec *r)
     r->per_dir_config = r->server->lookup_defaults;
 }
 
+typedef enum {
+    http_error_none,
+    http_error_badprotocol,
+    http_error_reject09,
+    http_error_badmethod09,
+} http_error;
+
+static http_error r_assign_protocol(request_rec *r,
+                                    const char *protocol,
+                                    int strict)
+{
+    int proto_num;
+    http_error error = http_error_none;
+
+    if (protocol[0] == 'H' && protocol[1] == 'T'
+            && protocol[2] == 'T' && protocol[3] == 'P'
+            && protocol[4] == '/' && apr_isdigit(protocol[5])
+            && protocol[6] == '.' && apr_isdigit(protocol[7])
+            && !protocol[8] && protocol[5] != '0') {
+        r->assbackwards = 0;
+        proto_num = HTTP_VERSION(protocol[5] - '0', protocol[7] - '0');
+    }
+    else if ((protocol[0] == 'H' || protocol[0] == 'h')
+                 && (protocol[1] == 'T' || protocol[1] == 't')
+                 && (protocol[2] == 'T' || protocol[2] == 't')
+                 && (protocol[3] == 'P' || protocol[3] == 'p')
+                 && protocol[4] == '/' && apr_isdigit(protocol[5])
+                 && protocol[6] == '.' && apr_isdigit(protocol[7])
+                 && !protocol[8] && protocol[5] != '0') {
+        r->assbackwards = 0;
+        proto_num = HTTP_VERSION(protocol[5] - '0', protocol[7] - '0');
+        if (strict && error == http_error_none)
+            error = http_error_badprotocol;
+        else
+            protocol = apr_psprintf(r->pool, "HTTP/%d.%d", HTTP_VERSION_MAJOR(proto_num),
+                                    HTTP_VERSION_MINOR(proto_num));
+    }
+    else if (protocol[0]) {
+        proto_num = HTTP_VERSION(0, 9);
+        if (error == http_error_none)
+            error = http_error_badprotocol;
+    }
+    else {
+        r->assbackwards = 1;
+        protocol = "HTTP/0.9";
+        proto_num = HTTP_VERSION(0, 9);
+    }
+    r->protocol = protocol;
+    r->proto_num = proto_num;
+    return error;
+}
+
 AP_DECLARE(int) ap_assign_request(request_rec *r,
                                   const char *method, const char *uri,
                                   const char *protocol)
 {
-    /* dummy, for now */
-    (void)r;
-    (void)method;
-    (void)uri;
-    (void)protocol;
+    core_server_config *conf = ap_get_core_module_config(r->server->module_config);
```

Comment:
Assign `method`, `uri` and `protocol` to a `request_rec`. These are coming from a received REQUEST bucket and need to be checked for validity. And if valid, all concerned fields need to be updated.

Again, code moved here from former http/1.x parsing.

Reply:
Maybe call it ap_assign_request_line()?

Reply:
🚴‍♀️Makes it easier to understand for all the h1 boomers, yes.

Reply:
done.